token 2022: add metadata pointer extension to js @solana/spl-token client

mistersimon opened the pull request:
Adding support for the metadata pointer extension


Split out of wider task to add metadata support for ease of review
https://github.com/solana-labs/solana-program-library/pull/5667

mistersimon commented on token/js/src/extensions/extensionType.ts:
```diff
@@ -8,29 +8,34 @@ import { DEFAULT_ACCOUNT_STATE_SIZE } from './defaultAccountState/index.js';
-    NonTransferableAccount,
-    TransferHook,
-    TransferHookAccount,
+    Uninitialized = 0,
```
There were some typescript errors if I created the unused `ConfidentialTransferFee`... instructions. Would need to define them in the below functions to work, and i'm not sure the approriate ones. 

The explicit definition is the same ones that Typescript automatically generates, so there is no change in the transplied javacript code.

buffalojoec replied:
Which TypeScript errors? I would much rather just add them to the enum and do nothing else, so that we don't have to track these numbers.

mistersimon replied:
I think theres two parts to this discussion.


1. Want to clarify that the numbers are actually used and need to match the values in rust, so they do need to be tracked. For example the below function relies on the numbers to unpack the extension data. I prefer it explicitly defined, its a good enforcement that they actually have meaning, but will defer to you. If we remove the numbers, we just need to make sure the the order matches the rust implementation, so the automatic values given by TS matches the rust.
https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/extensions/extensionType.ts#L189

1. The reason that unimplemented extensions like `ConfidentialTransferFee` are commented out, is that if they are defined it needs to be added to this function https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/extensions/extensionType.ts#L130C33-L130C33. Theres no default case, so typescript will complain because the function can return  undefined. While its an easy fix to add a default cause to throw an error - i thought it would be better for whoever is implementing the extension to enable it and configure it correctly. Happy to add a default case if preferred.

buffalojoec replied:
> 1. If we remove the numbers, we just need to make sure the the order matches the rust implementation

Yes, this is what we've been doing.

> 2\. While its an easy fix to add a default cause to throw an error

Yes do this! If the library doesn't support the extension yet, we just need to mark it as "unimplemented", which we can temporarily do with an error. Just so that it doesn't block progress on this PR.

mistersimon replied:
Okay, good that we are on the same page. I've gone ahead and removed it. 

However for 2 - it wasn't as easy as adding default error. This test case requires every extension to be either a Mint or Account extension, and i'm not confident which ones they should be.
https://github.com/solana-labs/solana-program-library/blob/master/token/js/test/unit/index.test.ts#L244

I removed all the numbering and only explicitly given a number to `MetadataPointer` as a work around. Take a look and hopefully its an acceptable compromise.

mistersimon reviewed (commented):
Hey @buffalojoec - I broke the metadata pointer extension into a separate PR. Thought it would be easier to review smaller changes.

mistersimon commented on token/js/src/extensions/extensionType.ts:
```diff
@@ -8,29 +8,34 @@ import { DEFAULT_ACCOUNT_STATE_SIZE } from './defaultAccountState/index.js';
+    // ConfidentialTransferFee = 16, // Not implemented yet
+    // WithdrawExcessLamports = 17, // Not implemented yet
+    MetadataPointer = 18,
+    // GroupPointer = 19, // Not implemented yet
```
TODO: Wrong numbers. Source of truth is here
https://github.com/solana-labs/solana-program-library/blob/master/token/program-2022/src/extension/mod.rs#L903

buffalojoec commented on token/js/src/extensions/metadataPointer/instructions.ts:
```diff
@@ -0,0 +1,95 @@
+    Update = 1,
+}
+
+export const initializeMetadataPointer = struct<{
```
nit: Let's call this something with `*Encoder` or `*Data`, just so it's marked as a serialization const. `initializeMetadataPointerData`?

mistersimon replied:
Updated!

buffalojoec commented on token/js/src/extensions/metadataPointer/instructions.ts:
```diff
@@ -0,0 +1,95 @@
+    authority: PublicKey;
+    metadataAddress: PublicKey;
+}>([
+    u8('instruction'), // prettier let me wrap please
```
At the top of the function block:

```javascript
// prettier-ignore
```

mistersimon replied:
Awesome, thats better. made the change

buffalojoec commented on token/js/src/extensions/metadataPointer/state.ts:
```diff
@@ -0,0 +1,41 @@
+export function getMetadataPointerState(mint: Mint): Partial<MetadataPointer> | null {
+    const extensionData = getExtensionData(ExtensionType.MetadataPointer, mint.tlvData);
+    if (extensionData !== null) {
+        const state: MetadataPointer = {};
+        const decoded = MetadataPointerLayout.decode(extensionData);
+
+        // Only add keys if they are defined (ignored zero public keys)
+        if (!decoded.authority.equals(PublicKey.default)) {
+            state.authority = decoded.authority;
+        }
+        if (!decoded.metadataAddress.equals(PublicKey.default)) {
+            state.metadataAddress = decoded.metadataAddress;
+        }
+
+        return state;
```
This is slick, but I think in this case it makes more sense to opt for `null`. That way, applications know they'll definitely have the fields defined, and `null` explicitly means the key is set to `None` (default).

mistersimon replied:
That makes sense. Made the change

buffalojoec reviewed (commented):
Nice! We are cruising right along with Token2022's JS support, thanks to you mate.

Just some small bits from my side, then we should be able to wrap soon.

mistersimon commented:
> Nice! We are cruising right along with Token2022's JS support, thanks to you mate.
> 
> Just some small bits from my side, then we should be able to wrap soon.

Thanks for the kind words @buffalojoec. 

Addressed most of your comments - wanted to clarify one thing on the enums before changing

buffalojoec commented on token/js/src/extensions/extensionType.ts:
```diff
@@ -31,6 +33,12 @@ export enum ExtensionType {
     NonTransferableAccount,
     TransferHook,
     TransferHookAccount,
+    // ConfidentialTransferFee, // Not implemented yet
+    // ConfidentialTransferFeeAmount, // Not implemented yet
+    MetadataPointer = 18, // Remove number once above extensions implemented
+    // TokenMetadata, // Not implemented yet
+    // GroupPointer, // Not implemented yet
+    // TokenGroup, //Not implemented yet
```
Yeah this works.

Tiny nit: There really isn't any reason to include any after `MetadataPointer` here. And on the next PR we just add `TokenMetadata`, no others.

buffalojoec replied:
```suggestion
    // ConfidentialTransferFee, // Not implemented yet
    // ConfidentialTransferFeeAmount, // Not implemented yet
    MetadataPointer = 18, // Remove number once above extensions implemented
```

mistersimon replied:
Makes sense. Changed!

buffalojoec commented on token/js/src/instructions/types.ts:
```diff
@@ -37,4 +37,8 @@ export enum TokenInstruction {
     CpiGuardExtension = 34,
     InitializePermanentDelegate = 35,
     TransferHookExtension = 36,
+    // ConfidentialTransferFeeExtension = 37,
+    // WithdrawalExcessLamports = 38,
+    MetadataPointerExtension = 39,
+    // GroupPointerExtension = 40,
```
Same tiny nit here on introducing trailing variants we don't need yet.

buffalojoec replied:
```suggestion
    // ConfidentialTransferFeeExtension = 37,
    // WithdrawalExcessLamports = 38,
    MetadataPointerExtension = 39,
```

mistersimon replied:
Makes sense. Changed!

buffalojoec reviewed (commented):
This looks great! Two little things I noticed.

mistersimon commented:
> This looks great! Two little things I noticed.

Made those changes. Let me know if there is anything else!

buffalojoec reviewed (approved):
Nice work!! JavaScript Token2022 support is shaping up nicely!
